hydromt/models/model.py

```diff
@@ -125,25 +133,47 @@ def __init__(
         self._forcing: Optional[XArrayDict] = None
         self._states: Optional[XArrayDict] = None
         self._results: Optional[XArrayDict] = None
-        # To be deprecated in future versions!
-        self._staticmaps = None
-        self._staticgeoms = None
 
         # file system
-        if root is None:
-            self.root: ModelRoot = ModelRoot(".", mode=mode)
-        else:
-            self.root: ModelRoot = ModelRoot(root, mode=mode)
+        self.root: ModelRoot = ModelRoot(root or ".", mode=mode)
 
-        self.region: ModelRegionComponent = ModelRegionComponent(self)
+        self._components: Dict[str, ModelComponent] = {}
+        self._add_components(components)
 
         self._defered_file_closes = []
 
         # model paths
         self._config_fn = self._CONF if config_fn is None else config_fn
-        self.logger.info(
-            f"Initializing {self._NAME} model from {dist_name} (v{version})."
-        )
+
+    def _add_components(self, components: dict[str, dict[str, Any]]) -> None:
+        """Add all components that are specified in the config file."""
+        for name, options in components.items():
+            type_name = options.pop("type")
+
+            self.add_component(
+                name, PLUGINS.component_plugins[type_name](self, **options)
+            )
+
+    def add_component(self, name: str, component: ModelComponent) -> None:
+        """Add a component to the model. Will raise an error if the component already exists."""
+        if name in self._components:
+            raise ValueError(f"Component {name} already exists in the model.")
+        if not name.isidentifier():
+            raise ValueError(f"Component name {name} is not a valid identifier.")
+        self._components[name] = component
+
+    def get_component(self, name: str, _: Type[T]) -> T:
+        """Get a component from the model. Will raise an error if the component does not exist."""
+        return cast(T, self._components[name])
+
+    def __getattr__(self, name: str) -> ModelComponent:
+        """Get a component from the model. Will raise an error if the component does not exist."""
+        return self._components[name]
+
+    @property
+    def region(self) -> ModelRegionComponent:
+        """Return the model region component."""
+        return self.get_component("region", ModelRegionComponent)
```

Comment:
Yes, because we ensure for the user that it is a `ModelRegionComponent`. Otherwise they will only get a `ModelComponent` and have to work dynamically.